---
output: html_notebook
author: "Ozan Berk Imir & Ernesto Segredo Otero"
date: "`r Sys.Date()`"
pdf_document: default
html_document: default
---

# Creation of an OrgDB from scratch using AnnotationForge (Version 3)

### Ozan Berk Imir and Ernesto Segredo Otero
### Last compiled on January 16th, 2025

We need a data frame that contain gene names, descriptions, relevant GO term associations and chromosome information. Such a database,
in complete form, did not exist for Candida albicans, so we will begin by merging two separate data frames with relevant information.
The data frames are as follows:

1. **uniprot**: Collects all protein functional information for all Candida albicans strains. Contains entry ID, protein and gene names.
2. **GO_data**: Collects all GO term associations for all Candida albicans strains. Contains entry ID, GO term, and evidence code.

We join these two data frames using the gene names from **uniprot** and the gene symbols from **GO_data**. Since the data frames are 
made up of many Candida albicans strains with many different taxonomy IDs, we have to pre-filter them after the merge, turn them into 
NAs and move on to remove the NAs after.

### Sources for the data

The data was sourced from the following:

1. [Uniprot](https://www.uniprot.org/)
2. [GO_data](http://www.candidagenome.org/download/go/)
3. Gene List: These genes are obtained from the results of a differential gene expression analysis, such as DESeq2.
4. **Optional:**[quickGO](https://www.ebi.ac.uk/QuickGO/) (to generate GO_data file for any organism of your choice)

**Note:** The GO_data association file can both be obtained from your organism-specific database or alternatively from the quickGO link 
above. The resource is easy to use and requires you to use the link and searching for your organism's tax ID to continue. 

The **Uniprot** data needs to be modified on the Uniprot website to create a data frame that contains the gene IDs, gene symbols, and
gene names. You could do so by (i) searching for the organism of interest ("Candida albicans" in this circumstance), (ii) clicking on
the "Customize columns" option, and (iii) selecting the columns of interest while removing all other columns. In this case:

1. Entry
2. Gene names
3. Gene names (primary)
4. Gene names (ordered locus)
5. Protein names
6. Reviewed
7. Entry name
8. Organism
9. Length

Once the data is retrieved from the sources, we can begin reading in the data and formatting our environment accordingly. 

# STEP 1: Create a merged data frame with all the necessary parameters listed above

1. Read the data once they have been downloaded from the sources.
2. Merge the two data frames using the gene names from the uniprot data frame and the gene symbols from the GO data frame.
3. Remove the gene IDs that belong to unwanted Candida albicans taxonomy IDs and turn them into NAs.
4. Remove all NAs from the merged_data data frame.

```{r}
library(tidyverse)
library(AnnotationForge)

# Read in the uniprot file
uniprot <- read.table("uniprotkb_candida_albicans_2024_07_16.tsv", sep = "\t", fill = TRUE, comment.char = "!", header = TRUE)

# Read the file, skipping the description lines that start with an exclamation mark
GO_data <- read.table("gene_association(1).cgd", sep = "\t", fill = TRUE, comment.char = "!", header = FALSE)

# Merge the two data frames
merged_data <- merge(uniprot, GO_data, by.x = "Gene.Names..primary.", by.y = "V3")
## Create a "gene" column that contains the genename synonymous to the gene ID from the geneset analysis
merged_data$gene <- sapply(merged_data$V11, function(x) {
  gsub("^(.{9}).*", "\\1", x)
})
## Remove the gene IDs that are from a different taxon (not in the right format) (turn them into NA)
merged_data$gene <- sapply(merged_data$gene, function(x) {
  ifelse(grepl("^.{2}[^_]", x), NA, x)
})
## Remove all NAs
merged_data <- merged_data[is.na(merged_data[,26])==FALSE,]
```


# STEP 2: Preparation of the Symbol, Chromosome, and GO data frames

From this point forward, we will follow the procedure outlined in this vignette here:
[AnnotationForge "Making Organism Packages" markdown](https://bioconductor.org/packages/release/bioc/vignettes/AnnotationForge/inst/doc/MakingNewOrganismPackages.html)

The OrgDB package is created from the three objects that contain the information listed:

### SYMBOL data frame structure

Contains the gene ID ("GID"), gene symbol ("SYMBOL"), and gene name ("GENENAME"). 

- The GID is the primary method of matching the gene information across all three data frames. It is acquired by isolating the first 9 
characters of the gene ID from the column with all gene information listed from the Gene Ontology ("GO") project. 
- The SYMBOL is the 3-4 character gene symbol that is used to identify the gene in the Candida albicans genome. Not all genes are given
a gene symbol, in which case the GID is assigned as the SYMBOL. 
- The GENENAME is the full gene name that is given to the gene.

We isolate the GID, SYMBOL and GENENAME columns (26th, 1st, and 5th columns respectively) and save it into **candidaSym**.

**candidaSym** then needs to be pre-processed such that:

1. No NA values exist in any column,
2. The columns are properly labeled as GID, SYMBOL and GENENAME,
3. No duplicate values are present in the SYMBOL and GENENAME columns,
4. The SYMBOL value is properly formatted (no more than 5 words are kept to describe function). This is done so future analysis can be
done so that specific gene functions can be extracted from genesets after over-representation analysis.

```{r}
# Now prepare the Symbol, Chromosome and GO data frames
candidaSym <- unique(merged_data[,c(26,1,5)])
candidaChr <- unique(merged_data[,c(26,19)])
candidaGO <- unique(merged_data[,c(26,13,15)])

# Pre-process the Symbol data frame -- DONE!
candidaSym <- candidaSym[is.na(candidaSym[,2])==FALSE,]
candidaSym <- candidaSym[is.na(candidaSym[,3])==FALSE,]
colnames(candidaSym) <- c("GID","SYMBOL","GENENAME")
candidaSym <- candidaSym[!duplicated(candidaSym$SYMBOL), ]
candidaSym <- candidaSym[!duplicated(candidaSym$GENENAME), ]
candidaSym <- candidaSym %>%
  mutate(GENENAME = sapply(strsplit(GENENAME, " "), function(x) {
    ## Keep the first 5 words
    first_five <- paste(x[1:min(5, length(x))], collapse = " ")
    ## Remove any text after the right-facing parenthesis
    gsub("\\(.*", "", first_five)
  }))
```

### CHROMOSOME data frame structure

Contains the gene ID("GID") and the chromosome number ("CHROMOSOME"). 

The CHROMOSOME information is extracted from the GID information as GID contains the chromosome number as well as the ORF number for a 
given gene. There are 8 pairs of chromosomes in the Candida albicans genome, 7 pairs of autosomal chromosomes and 1 R chromosome pair.

**candidaChr** needs to be pre-processed such that:

1. No NA values exist in any column,
2. The columns are properly labeled as GID and CHROMOSOME,
3. The CHROMOSOME values are properly formatted (only numbers and the letter "R" are kept).

```{r}
# Pre-process the Chromosome data frame -- DONE!
candidaChr <- candidaChr[is.na(candidaChr[,2])==FALSE,]
colnames(candidaChr) <- c("GID","CHROMOSOME")
candidaChr$CHROMOSOME <- substr(candidaChr$CHROMOSOME,2,2)
candidaChr$CHROMOSOME <- sapply(candidaChr$CHROMOSOME, function(x) {
  gsub("[^0-9R]", NA, x)
})
candidaChr <- candidaChr[is.na(candidaChr$CHROMOSOME)==FALSE,]
```

### GO data frame structure

Contains the gene ID("GID"), the GO term ("GO"), and the evidence code ("EVIDENCE"). 

**candida GO** needs to be pre-processed such that:

1. No NA values exist in any column,
2. The columns are properly labeled as GID, GO, and EVIDENCE.

```{r}
# Pre-process the GO data frame -- DONE!
candidaGO <- candidaGO[is.na(candidaGO[,2])==FALSE,]
candidaGO <- candidaGO[is.na(candidaGO[,3])==FALSE,]
colnames(candidaGO) <- c("GID","GO","EVIDENCE")
```

At this stage, you can switch the bottom OrgDB object parameters between the Candida albicans and the Mustela putorius data, depending
on your organism of interest.

# STEP 3: Creation of the OrgDB package

Now that we have the Symbol, Chromosome, and GO data frames ready, we can create the OrgDB package using the **makeOrgPackage** 
function which will take all annotations, and will combine them in an R package that can be used for downstream analysis. In creating
your OrgDB package, you will need to specify the following parameters:

1. **gene_info**: The gene information data frame.
2. **version**: The version of the OrgDB package. Any number you are on with the package works.
3. **maintainer**: Your information with this format: "Name <email>" with correct bracket placement (don't change format).
4. **author**: Same thing as maintainer (don't change format).
5. **outputDir**: The directory where the OrgDB package will be saved. Can just leave as "." and it will be added to WorkingDirectory.
6. **tax_id**: The taxonomy ID of the organism you are working with. For Candida albicans, it is 237561.
7. **genus**: The genus of the organism you are working with.
8. **species**: The species of the organism you are working with.
9. **goTable**: What type of analysis you are going to perform with package. Could be replaced with ENSEMBL or other means of analysis.

The code block below may only be run ONCE, and then you should COMMENT OUT that code block, or you will face an error. This happens
because you already have an Org Package present in your working directory, and the solution that R suggests gives an error as well.

```{r}
# COMMENT OUT BELOW CODE AFTER RUNNING IT ONCE!!!
# Call the function
# makeOrgPackage(gene_info=candidaSym, chromosome=candidaChr, go=candidaGO,
#                version="0.1",
#                maintainer="Ozan Imir <obi203@nyu.edu>",
#                author="Ozan Imir <obi203@nyu.edu>",
#                outputDir = ".",
#                tax_id="237561",
#                genus="Candida",
#                species="albicans",
#                goTable="go")

# --- BLOCK OUT UP TO THIS LINE ---

# Keep the code below in order to install the created OrgDB package, DO NOT COMMENT OUT BELOW!!!
# Install the Candida albicans OrgDB package as a local package before loading it to your local environment
install.packages("./org.Calbicans.eg.db", type = "source", repos=NULL)
```

# STEP 4: Prepare the packages, directories and DESeq2 lists for Overrepresentation Analysis

We will use the **clusterProfiler** package to perform an overrepresentation analysis on the genes that we got from a DESeq2 pipeline.
The gene list used in this case should contain the same gene symbols as the ones that have been fed to the OrgDB package as SYMBOL.

To test the efficacy of the orgDB database, we will perform an overrepresentation analysis using the **topGO** package utilizing the
vignette designed by Mohammad Khalfan here:
[Over-Representation Analysis with ClusterProfiler](https://learn.gencore.bio.nyu.edu/rna-seq-analysis/over-representation-analysis/)

### Register the directories and install / load the packages

We use the following packages in the over-representation analysis:

1. **clusterProfiler**: Contains the main functions that undertake the over-representation analysis.
2. **pathview**: Contains the functions that allow for the visualization and integration of the over-representation analysis.
3. **wordcloud**: Contains the functions that allow for the creation of a word cloud.
4. **org.Calbicans.eg.db**: The OrgDB package that we created in the previous steps.

```{r, message=F, warning=F}
# Indicate the filepaths for the subdirectories
data_filepath <- "data/"
geneset_filepath <- paste0(data_filepath, "time_72_vs_2_results.csv")

# Install the two commented out packages only once if you haven't installed them before, then comment out
# BiocManager::install("clusterProfiler", version = "3.8")
# BiocManager::install("pathview")
# install.packages("wordcloud")
library(clusterProfiler)
library(pathview)
library(wordcloud)

# Load in the Candida albicans OrgDB object created in the GO-term pipeline
library(org.Calbicans.eg.db)
```

### Prepare Input

```{r}
cluster1_filepath <- paste0(data_filepath, "cluster_1_genes.txt")
cluster2_filepath <- paste0(data_filepath, "cluster_2_genes.txt")
cluster3_filepath <- paste0(data_filepath, "cluster_3_genes.txt")
cluster4_filepath <- paste0(data_filepath, "cluster_4_genes.txt")
cluster5_filepath <- paste0(data_filepath, "cluster_5_genes.txt")
cluster6_filepath <- paste0(data_filepath, "cluster_6_genes.txt")
cluster7_filepath <- paste0(data_filepath, "cluster_7_genes.txt")
cluster8_filepath <- paste0(data_filepath, "cluster_8_genes.txt")

# Read in the gene lists
cluster1 <- readLines(cluster1_filepath)
cluster2 <- readLines(cluster2_filepath)
cluster3 <- readLines(cluster3_filepath)
cluster4 <- readLines(cluster4_filepath)
cluster5 <- readLines(cluster5_filepath)
cluster6 <- readLines(cluster6_filepath)
cluster7 <- readLines(cluster7_filepath)
cluster8 <- readLines(cluster8_filepath)
```

# STEP 5: Over-representation Analysis

The Over-representation Analysis aims to identify the biological pathways that are significantly enriched in the gene list that we have
obtained from the DESeq2 analysis in comparison to their abundance in the genome.

### Creation of the enrichGO object

*Parameters*:

**Ontology** Options: ["BP", "MF", "CC"]
**keyType** This is the source of the annotation (gene ids). The options vary for each annotation. In the example of *org.Dm.eg.db*, 
the options are:

"CHROMOSOME"  "EVIDENCE"    "EVIDENCEALL" "GENENAME"    "GID"         
"GO"          "GOALL"       "ONTOLOGY"    "ONTOLOGYALL" "SYMBOL"
  
Check which options are available with the `keytypes` command, for example `keytypes(org.Dm.eg.db)`.

### Create the object

```{r}
# reading in input from deseq2
df = read.csv(geneset_filepath, header=TRUE)

# we want the log2 fold change 
original_gene_list <- df$log2FoldChange

# name the vector
names(original_gene_list) <- df$Locus

# omit any NA values 
gene_list <- na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list <- sort(gene_list, decreasing = TRUE)

go_enrich_cluster1 <- enrichGO(gene = cluster1,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)

go_enrich_cluster2 <- enrichGO(gene = cluster2,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)

go_enrich_cluster3 <- enrichGO(gene = cluster3,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)

go_enrich_cluster4 <- enrichGO(gene = cluster4,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)

go_enrich_cluster5 <- enrichGO(gene = cluster5,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)

go_enrich_cluster6 <- enrichGO(gene = cluster6,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)

go_enrich_cluster7 <- enrichGO(gene = cluster7,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)

go_enrich_cluster8 <- enrichGO(gene = cluster8,
                      universe = names(gene_list),
                      OrgDb = "org.Calbicans.eg.db", 
                      keyType = 'GID',
                      readable = TRUE,
                      ont = "ALL",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)
```

### Table of results

```{r}
head(go_enrich_cluster1)
```

### Dotplot

```{r echo=TRUE}
dotplot(go_enrich_cluster1)
ggsave("dotplot_cluster1.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
dotplot(go_enrich_cluster2)
ggsave("dotplot_cluster2.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
dotplot(go_enrich_cluster3)
ggsave("dotplot_cluster3.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
dotplot(go_enrich_cluster4)
ggsave("dotplot_cluster4.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
dotplot(go_enrich_cluster5)
ggsave("dotplot_cluster5.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
dotplot(go_enrich_cluster6)
ggsave("dotplot_cluster6.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
dotplot(go_enrich_cluster7)
ggsave("dotplot_cluster7.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
dotplot(go_enrich_cluster8)
ggsave("dotplot_cluster8.png", plot = last_plot(), width = 10, height = 8, dpi = 300)
```

# Summary

```{r}
# Print session info
sessioninfo::session_info()
```